### ISSUE TITLE
Flattening parameters

### ISSUE THREAD

**Issue by MilesCranmer** [google/flax#5]
Hi,

Great package! I like the syntax.

Is there an easy way to pack and unpack parameters for a `flax.nn.Module`? I saw this in the optimizer code:

```
    new_params = jax.tree_unflatten(treedef, new_params_flat)
    new_param_states = jax.tree_unflatten(treedef, new_states_flat)
```
as well as this:
```
  def init_state(self, params):
    sub_states = []
    for traversal, opt in zip(self.traversals, self.sub_optimizers):
      params_t = list(traversal.iterate(params))
      state = opt.init_state(params_t)
      sub_states.append(state)
```
but was wondering if there was a simpler way to convert the parameters to a single array (which I can still compute gradients on)? I'm interested in doing some operations on the parameters as a single vector as part of the loss function.

Thanks,
Miles

**Comment by avital** [comment]
Hi Miles! We are still working on the best patterns for parameter serialization/linearization. We're focusing on getting our documentation clearer and completing our "open source transition" first before we're ready to engage with new users, but let's keep this issue open and we'll get back to you once we have a clear pattern in place!

**Comment by avital** [comment]
Hi @MilesCranmer, sorry for the delay -- we're ramping up for a more official open source release now so going through all of the open issues again.

Can you clarify the question? Indeed https://jax.readthedocs.io/en/latest/jax.tree_util.html#jax.tree_util.tree_flatten an the rest of `jax.tree_util` are the way to flatten parameters into a flat array (though it's still an array of jax.numpy.arrays, not a single flattened value -- this is important to keep the runtime performance of XLA/JAX).

Then you should be able to compute gradients with `jax.grad` yourself, e.g. the following code snippet (assuming `loss_fn` is already defined):

```py
treedef = jax.tree_util.tree_structure(model.params)
def loss_fn_flat(flat_params):
  unflat_params = jax.tree_util.tree_unflatten(treedef, flat_params)
  return loss_fn(unflat_params)

flat_params = jax.tree_util.tree_leaves(model.params)
jax.grad(loss_fn_flat)(flat_params)
```

If what you want is to /fully/ flatten parameters into a single array of scalars, you'll have to do a bit more work but more importantly JAX will no longer be able to store parameters directly on device, so training will be much slower.

**Comment by marcvanzee** [comment]
Closing this since the question seems to have been answered by Avital 21 days ago. Miles, if think are still unclear for you, please feel free to re-open this issue!

**Comment by ddrous** [comment]
> If what you want is to /fully/ flatten parameters into a single array of scalars, you'll have to do a bit more work but more importantly JAX will no longer be able to store parameters directly on device, so training will be much slower.

Hi @avital, is this still the case? And could you please explain what you mean by "Jax will no longer be able to store parameters directly on device" ? Thank you :)

**Comment by cgarciae** [comment]
Hey @ddrous, stacking all parameters into a single flat array will probably lead to OOM in practice (for big enough models) as you are both duplicating memory the amount of memory needed and requiring that a large amount of memory exists contiguously.

**Comment by ddrous** [comment]
Thanks for such an elegant answer @cgarciae. I'm doing sensitivity analysis and I need all my weights completely flattened. I don't suppose there's a better way to do this while avoiding (some of) the issues you outlined ?

**Comment by cgarciae** [comment]
Maybe for a single sample this is not a problem. Are you having issues?

**Comment by ddrous** [comment]
No issues yet. But I'm afraid it's going to become intractable when I scale it to deeper neural networks and large dataset.
